Disallow functions in react-redux#connect()

mcsf opened the pull request:
See https://github.com/Automattic/wp-calypso/issues/14024

My first eslint rule, so be gentle. 😄

mcsf committed:
Disallow functions in react-redux#connect()

mcsf commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -0,0 +1,65 @@
+/**
+ * @fileoverview Disallow the use of this.translate
```
Oopsies, forgot this

aduth commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -0,0 +1,65 @@
+
+function isStateInArgs( node ) {
+	return node.params.some( function( paramNode ) {
+		return paramNode.name === 'state';
```
Matching by assigned name could have some false positives, but the alternative would be pretty cumbersome I think (tracking assigned `connect` named import to its first function argument when invoked). Wondering if people would use [React's `this.setState`](https://facebook.github.io/react/docs/react-component.html#setstate) function callback form with `state` as the argument name.

mcsf replied:
Good point with `setState`. I've just pushed dc0aecf to address that.

aduth commented on docs/rules/redux-no-bound-selectors.md:
````diff
@@ -0,0 +1,20 @@
+  getSite: ( id ) => getSite( state, id ),
+} );
+connect( mapState )( MyComponent );
+```
````
Are there any good "not warnings" examples we could include (as per convention with these docs) that might illustrate good alternatives?

aduth reviewed (commented):
Lodash's `partial` could be similarly abused, though dunno that we care to address it since there's always going to be workarounds that we can't anticipate.

mcsf committed:
Don't report instances within `this.setState`

mcsf commented:
Back from the meetup. 😄 

Another thing this catches is legitimate uses where state is recombined in a selector. Would you agree if we only reported a node if it is not wrapped in a `createSelector` call? Example caught by this rule:

```js
export const getSiteCollisions = createSelector(
	( state ) => {
		return map( filter( state.sites.items, ( wpcomSite ) => {
			const wpcomSiteUrlSansProtocol = withoutHttp( wpcomSite.URL );
			return ! wpcomSite.jetpack && some( state.sites.items, ( jetpackSite ) => {
				return (
					jetpackSite.jetpack &&
					wpcomSiteUrlSansProtocol === withoutHttp( jetpackSite.URL )
				);
			} );
		} ), 'ID' );
	},
	( state ) => state.sites.items
);
```

That way we also make sure that, aside from dodging the impact of repeated function binding via `connect`, we also recycle object references thanks to caching.

aduth commented:
Rather than working to exempt cases, can we improve the intended targeting by testing only when bind occurs in context of `connect` ?

mcsf commented:
> Rather than working to exempt cases, can we improve the intended targeting by testing only when bind occurs in context of connect ?

How would you deal with the common pattern of defining `mapState` as a
standalone/module function and passing it to `connect`? A simple test for a
couple of identifiers (e.g. `mapState`, `mapStateToProps`)? Or would you
attempt something more sophisticated, like tracking the variable passed to
`connect` and analyzing the corresponding function?

aduth commented:
>Or would you attempt something more sophisticated, like tracking the variable passed to `connect` and analyzing the corresponding function?

This would seem to be best so far as limiting the number of false positives/negatives, but also the most difficult to implement.

gwwar commented:
Just thinking of ideas, but it looks like we have access to the full ast tree, so we might break it down by:
- Verifying that we have a connect import from `react-redux`. And keeping a note of it's local name. Though it'll probably just  be `connect` in most cases.
- Find the `connect` callExpression
- Get the first argument
- If it's a function, run the checks
- if it's an identifier search for it in the tree and run the checks

mcsf committed:
Look for invocations of `connect` to inspect `mapState`

- Instead of operating on any function that has `state` as an argument,
  only consider those that are actually passed as the first argument to
  `connect` — whether as an anonymous function or as a variable.

- Longer comments in the source should guide the reader through the
  implementation.

mcsf commented:
Alright, you two — I went with the longer route in 91596de. Added a test file showcasing examples. Let me know what you think! @gwwar @aduth

gwwar commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -15,59 +15,152 @@ const BIND_ERROR_MESSAGE = "Don't bind functions within `connect`." + REF;
+			},
+
+			CallExpression: function( node ) {
+				if ( node.callee.name !== 'connect' ) {
```
Should we also check in `ImportDeclaration` that we have a connect import from `react-redux`?

gwwar replied:
Would it be worth walking `context.getSourceCode()` to see if we have a matching import? I think we'd be able to exit early in these handlers.

gwwar replied:
Doing a quick search in Calypso, I think we can assume that it's probably from react-redux.

mcsf replied:
> I think we can assume that it's probably from react-redux.

That was my thinking. If anyone ever does use something else called `connect`, they/we can deal with it accordingly.

gwwar commented:
Overall this appears to work really well. I'll let @aduth 👀  to see if any of this can be simplified. Besides that we'll want to 
- update the readme, to add some bad examples
- bump the version number
- update the main readme to point at the new rule
- update the changelog

mcsf committed:
Update rule documentation

mcsf committed:
Update all the updates!

aduth committed:
Test connect with no arguments as valid

aduth commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -0,0 +1,166 @@
+
+				let mapStateBody;
+
+				if ( mapStateNode.type === 'Identifier' ) {
```
Pushed a failing test caused by assumption of arguments at this line.

aduth commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -0,0 +1,166 @@
+				}
+
+				if ( mapStateBody ) {
+					nodesToReportOn.push( mapStateBody );
```
~~Why do we wait 'til `Program:exit` to report on the node?~~

__Edit:__ Hmm, we need the whole set to know whether it's an issue? _Continues digesting_ 😄 
__Edit2:__ Yeah, disregard this one.

aduth commented on lib/rules/redux-no-bound-selectors.js:
```diff
@@ -0,0 +1,166 @@
+				}
+			},
+
+			Identifier: onIdentifierEnter,
```
I'm a little worried about the performance implications of an Identifier visitor. Do you think `CallExpression.callee.property.name` might be reasonable to check (and potentially more accurate) instead?

![image](https://cloud.githubusercontent.com/assets/1779930/26510873/42008f76-422d-11e7-8b56-3f093726413c.png)

http://astexplorer.net/#/gist/1ce6f31efbd822108c1227fbb336098a/7c73e810541bec4f0f94dd486de21f49198ca697

aduth committed:
Test bind as valid property access

aduth commented:
Pushed another failing test relating to https://github.com/Automattic/eslint-plugin-wpcalypso/pull/38#discussion_r118779849 . Maybe a stretch, but testing any `bind` identifiers within `connect` seems a bit too overreaching.

aduth committed:
Make failing bind identifier case a little less function-like

mcsf commented:
Thanks for the thorough input, @aduth. (I'll probably only work on this on Monday)

mcsf committed:
Guard against 0-arg `connect()` calls

mcsf committed:
Add failing tests for lodash bind, partial, partialRight

mcsf committed:
Ditch Identifier rule, use CallExpression, catch more binding cases

mcsf commented:
Feedback addressed, with tests improved and satisfied. Think it's 🚢-able? @gwwar @aduth

mcsf committed:
Update docs to mention lodash#partial*

mcsf commented:
I've been catching (and hopefully preventing, but only out of luck by stumbling upon them) a couple of new occurrences of the anti-pattern in Calypso. Let's get this merged! :)

gwwar commented:
I'm 👍  with the changes here. @aduth did you have any additional feedback?

aduth reviewed (approved):
To be clear, these two are valid issues? They're flagged for me in running the rule on Calypso's codebase. I'm guessing we're encouraging to create external selectors than mapping state within `connect`'s `mapStateToProps` ?

https://github.com/Automattic/wp-calypso/blob/f75c5ee/client/my-sites/theme/theme-features-card.js#L45
https://github.com/Automattic/wp-calypso/blob/f75c5ee/client/my-sites/themes/theme-options.js#L181-L185

Otherwise LGTM 👍

mcsf commented:
Correct, @aduth, to me it would make sense to refactor those with memoized selectors (cc @ockham). Many thanks for the reviews, you two!

mcsf commented:
(Aw, since this was my first merge in this repo the preset was to merge without squashing :/ sorry about that.)
